daemon/cmd/daemon.go

```diff
@@ -582,8 +582,80 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	debug.RegisterStatusObject("ongoing-endpoint-creations", d.endpointCreations)
 
 	d.k8sWatcher.RunK8sServiceHandler()
+
+	if option.Config.DNSPolicyUnloadOnShutdown {
+		log.Debugf("Registering cleanup function to unload DNS policies due to --%s", option.DNSPolicyUnloadOnShutdown)
+
+		// add to pre-cleanup funcs because this needs to run on graceful shutdown, but
+		// before the relevant subystems are being shut down.
+		cleaner.preCleanupFuncs.Add(func() {
+			// Stop k8s watchers
+			log.Info("Stopping k8s service handler")
+			d.k8sWatcher.StopK8sServiceHandler()
+
+			// Iterate over the policy repository and remove L7 DNS part
+			needsPolicyRegen := false
+			removeL7DNSRules := func(pr policyAPI.Ports) error {
+				portProtocols := pr.GetPortProtocols()
+				if len(portProtocols) == 0 {
+					return nil
+				}
+				portRule := pr.GetPortRule()
+				if portRule == nil || portRule.Rules == nil {
+					return nil
+				}
+				dnsRules := portRule.Rules.DNS
+				log.Debugf("Found egress L7 DNS rules (portProtocol %#v): %#v", portProtocols[0], dnsRules)
+
+				// For security reasons, the L7 DNS policy must be a
+				// wildcard in order to trigger this logic.
+				// Otherwise we could invalidate the L7 security
+				// rules. This means if any of the DNS L7 rules
+				// have a matchPattern of * then it is OK to delete
+				// the L7 portion of those rules.
+				hasWildcard := false
+				for _, dns := range dnsRules {
+					if dns.MatchPattern == "*" {
+						hasWildcard = true
+						break
+					}
+				}
+				if hasWildcard {
+					portRule.Rules = nil
+					needsPolicyRegen = true
+				}
+				return nil
+			}
+
+			policyRepo := d.GetPolicyRepository()
+			policyRepo.Mutex.Lock()
+			policyRepo.IterateLocked(func(rule *policyAPI.Rule) {
```

Comment:
In Cilium we read this roughly as `IterateWhileLocked()`: The idea is that everything should be more-or-less 'safe' when using the regular version (eg `Iterate()`). But if you need to hold the lock across multiple operations for some reason, then expose a special locked version that informs the caller that they need to be considering the correct locking to ensure the safety.

Reply:
Is it not more effective to have `IterateLocked` do the locking? or is the idea that the caller may want to perform some other actions once the tight loop returns?

Reply:
Usually when we expose functions like `IterateLocked()`, the caller wants to do other operations or do multiple operations while holding the lock. Looking at this case though, I think it would be OK to move the locking inside and rename `IterateLocked()` -> `Iterate()`.